doc/man/man4/nsdev.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/nsdev.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH NSDEV 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B nsdev
\- named \fISTREAMS\fP device
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/stropts.h>
.HP
.BI "int " fd " = open(" \*(lq/dev/xxx.yyy\*(rq ", " oflags );
.\"
.\"
.SH DESCRIPTION
.PP
.B nsdev
is a \*(lqNamed
.I STREAMS
Device\*(rq pseudo-device driver similar in some
respects to
.BR clone (4).
.B nsdev
is perhaps the only assigned major device number that is ever needed by a
.I STREAMS
device driver.
.PP
Whereas the
.BR clone (4)
pseudo-device driver transforms the device number
.I dev
into
.BI makedevice(getminor( dev "), " 0 ),
the
.B nsdev
pseudo-device driver transforms the device number
.I dev
into
.BI makedevice( major ", getminor(" dev )),
where
.I major
is determined from the name of the device rather than the device number.
.PP
To determine the
.I major
component of the device number,
.B nsdev
looks for a registered
.I STREAMS
device with the same initial substring (before a dot) as the name of the device opened
.RI ( xxx
in the
.RI \(lq SYNOPSIS \(rq
above).
.PP
If no registered
.I STREAMS
device can be found that match the initial substring,
.IR xxx ,
of the device name, the
.B nsdev
requests that the kernel load the module named
.BI streams- xxx\fR,
where
.I xxx
is the longest initial substring of the device name that does not contain a
period (\(lq.\(rq) character.  This makes it possible to create Named
.I STREAMS
Devices knowing only the character major device number associated with the
.B nsdev
driver and demand load other
.I STREAMS
modules and drivers that may then dynamically allocate their major device
numbers (or not allocate any major device numbers at all).
.PP
For additional information on device numbers, see the
.RI \(lq "Device Numbering" \(rq
section of the
.BR "Linux Fast-STREAMS Programmer's Guide" ,
.BR SPG (9),
or the
.BR dev_t (9)
manual page.
.\"
.SS
Super Cloning
.PP
The
.B nsdev
supports the concept of super cloning when used in conjunction with the
traditional
.BR clone (4)
device.
A super clone device has the clone device major device number as a major
number and the
.B nsdev
major device number as a minor number.  This is a traditional clone device
application to the
.B nsdev
pseudo-device driver.
This arrangement has the characteristic that when such a
.BR clone (4)
device is opened, it chains the open call to the
.B nsdev
with the
.B CLONEOPEN
flag set in
.IR sflag .
The Named
.I STREAMS
Device then chains the open to the device number derived
from the device name and the minor device number (in this case zero (0)) as
normal.
.\"
.SS
Named Clone Devices
.PP
The
.B nsdev
can be used to create named clone devices.  A named clone device can be
created by assigning the
.B ndev
major device number to devices with the name
.BI clone. yyy
and assigning the traditional clone device minor number (see
.BR clone (4)).
In this way, it is not even necessary to know the device major number of the
.BR clone (4)
device.
.\"
.\"
.SH NOTICES
.PP
Performing an
.BR open (2s)
of the
.B nsdev
can result in the loading of (or an attempt to load) the kernel module
named
.BI streams- xxx\fR.
.PP
If the device opened is a symbolic link, the name of the actual file is the
name used for determining the
.I major
device number of the cloned device.
If the device opened is a hard link, the name of the opened file is the name
used for determining the
.I major
device number of the cloned device.
.PP
.B nsdev
is used by the
.B Linux Fast-STREAMS
.[
lfs
.]
system provided drivers to limit the use of major device numbers to a single
major device number for the
.I STREAMS
system installed device drivers.
Clonable minor devices are allocated against a single
.I STREAMS
system major device number rather than allocating new major device numbers for
each facility.
.PP
When
.B nsdev
loads, it assigns a major device number either statically or dynamically.  In
either event, the major device number associated with the
.B nsdev
can be found by search the
.B /proc/devices
file for the major device number assigned to
.RB \*(lq nsdev \*(rq.
.\"
.\"
.SH IMPLEMENTATION
.PP
.B nsdev
is not implemented as a normal
.I STREAMS
driver: it is implemented as a shadow special device node under
.BR "Linux Fast-STREAMS" .
.[
lfs
.]
.B Linux Fast-STREAMS
shadow special device nodes call
.BR nsdev_open ()
from
.BR cdev_open ()
instead of
.BR spec_open ()
permitting the
.B nsdev
device to intercept and redirect the open call to the appropriate device.
This is consitent with the descriptions for the internal
.BR clone_open ()
under
.IR "SVR 4" .
.[
magic
.]
.\"
.\"
.SH EXAMPLES
.PP
Following are some examples of how to use the
Named
.I STREAMS
Device:
.SS
Named
.I STREAMS
Device Example
.PP
If the following is performed to create the device (assuming that major device
number 56 is the major device number of the Named
.I STREAMS
Device):
.sp
.nf
\fC\s-1\
$> mknod -m 0666 /dev/mystream.04 c 56 3
$> ln -s /dev/mystream.04 /dev/ticlts
\s+1\fP
.fi
.sp
then opening
.RB \*(lq /dev/ticlts \*(rq
will result in opening minor device number
.I 3
of the
.RB \*(lq mystream \*(rq
device.
The device will be opened with the
.I sflag
argument to
.BR qopen (9)
set to
.B DRVOPEN
unless the device has had cloning set using
.BR D_CLONE
during registration of the major or minor device nodes (see
.BR register_strdev (9)),
or unless
.BR SAP_CLONE
was set using the
.I STREAMS
Administrative Driver
.BR sad (4)
for the ultimate device node.  In either of these later cases, the device will
be opened with a
.I sflag
argument of
.BR CLONEOPEN .
.SS
Super Clone Example
.PP
If the following is performed to create a
.BR clone (4)
device in addition to the above example (assuming that the
.BR clone (4)
major device number is 74):
.sp
.nf
\fC\s-1\
$> mknod -m 0666 /dev/mystream.00 c 56 0
$> mknod -m 0666 /dev/mystream c 74 56
$> ln -s /dev/mystream my_super_clone_device
\s+1\fP
.fi
.sp
then opening either
.RB \*(lq my_super_clone_device \*(rq
or
.RB \*(lq /dev/mystream \*(rq
will result in opening minor device number
.I 0
of the
.RB \*(lq mystream \*(rq
device with the
.I sflag
argument to
.BR qopen (9)
being set to
.BR CLONEOPEN .
For additional information on the super cloning technique using the Named
.I STREAMS
Device, see
.BR clone (4).
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/nsdev
The external filesystem named
.I STREAMS
device.
.TP
.BR /dev/streams/nsdev ", " /devices/nsdev
The
.BR specfs (5)
shadow special filesystem named
.I STREAMS
device.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR specfs (5),
.BR open (2s),
.BR qopen (9),
.BR register_strdev (9),
.BR clone (4),
.BR sad (4),
.BR SPG (9),
.BR dev_t (9),
.BR STREAMS (9).
.\"
.\"
.SH BUGS
.PP
.B nsdev
has no known bugs.
.\"
.\"
.SH COMPATIBLITY
.PP
.B nsdev
is compatible with
.I SVR 4.2
.[
svr42
.]
and implementations based on
.IR "SVR 4.2" ,
including
.IR AIX \(rg,
.IR HP-UX \(rg,
.IR OSF/1 \(rg,
.IR Solaris \(rg,
.IR SUPER-UX \(rg,
.IR UXP/V \(rg,
with the following compatibility considerations:
.IP \(em \w'\(em\(em'u
.B nsdev
is a
.B Linux Fast-STREAMS
specific device.
.IP \(em
Although
.B nsdev
is
.B Linux Fast-STREAMS
specific, the operation of this device is a file system artifice and is
transparent to both the
.I STREAMS
device user and the
.I STREAMS
device driver itself.
.IP \(em
When the
.B nsdev
is opened directly, the driver has no indication that it has been opened via
the Named
.I STREAMS
Device and, for all intents and purposes, appears to the
driver as having been opened as a direct minor node of the driver.
.PP
See
.BR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
.B nsdev
is a
.B Linux Fast-STREAMS
specific mechanism.
Conformance is verified using the
.BR test-nsdev (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
.B nsdev
first appeared in
.BR "Linux Fast-STREAMS" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/nsdev.c
/*****************************************************************************

 @(#) File: src/drivers/nsdev.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/nsdev.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#ifdef NEED_LINUX_AUTOCONF_H
#include NEED_LINUX_AUTOCONF_H
#endif
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#if defined(CONFIG_KMOD) || defined(CONFIG_MODULES)
#include <linux/kmod.h>
#endif

#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>

#include "sys/config.h"
#include "src/modules/sth.h"

#define NSDEV_DESCRIP	"STREAMS Named Stream Device (NSDEV) STREAMS Driver"
#define NSDEV_EXTRA	"Part of Linux Fast-STREAMS"
#define NSDEV_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define NSDEV_REVISION	"OpenSS7 src/drivers/nsdev.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define NSDEV_DEVICE	"SVR 4.2 MP STREAMS Named Stream Device (NSDEV) Driver"
#define NSDEV_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define NSDEV_LICENSE	"GPL"
#define NSDEV_BANNER	NSDEV_DESCRIP	"\n" \
			NSDEV_EXTRA	"\n" \
			NSDEV_COPYRIGHT	"\n" \
			NSDEV_REVISION	"\n" \
			NSDEV_DEVICE	"\n" \
			NSDEV_CONTACT	"\n"
#define NSDEV_SPLASH	NSDEV_DEVICE	" - " \
			NSDEV_REVISION	"\n"

#ifdef CONFIG_STREAMS_NSDEV_MODULE
MODULE_AUTHOR(NSDEV_CONTACT);
MODULE_DESCRIPTION(NSDEV_DESCRIP);
MODULE_SUPPORTED_DEVICE(NSDEV_DEVICE);
MODULE_LICENSE(NSDEV_LICENSE);
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_NSDEV_MODULE */

#ifndef CONFIG_STREAMS_NSDEV_NAME
//#define CONFIG_STREAMS_NSDEV_NAME "nsdev"
#error "CONFIG_STREAMS_NSDEV_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_NSDEV_MODID
//#define CONFIG_STREAMS_NSDEV_MODID 7
#error "CONFIG_STREAMS_NSDEV_MODID must be defined."
#endif
#ifndef CONFIG_STREAMS_NSDEV_MAJOR
//#define CONFIG_STREAMS_NSDEV_MAJOR 0
#error "CONFIG_STREAMS_NSDEV_MAJOR must be defined."
#endif

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-nsdev");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

#ifndef CONFIG_STREAMS_NSDEV_MODULE
static
#endif
modID_t modid = CONFIG_STREAMS_NSDEV_MODID;

#ifdef CONFIG_STREAMS_NSDEV_MODULE
#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module id number for NSDEV driver.");
#endif				/* CONFIG_STREAMS_NSDEV_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_NSDEV_MODID));
MODULE_ALIAS("streams-driver-nsdev");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

#ifndef CONFIG_STREAMS_NSDEV_MODULE
static
#endif
major_t major = CONFIG_STREAMS_NSDEV_MAJOR;

#ifdef CONFIG_STREAMS_NSDEV_MODULE
#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for NSDEV driver.");
#endif				/* CONFIG_STREAMS_NSDEV_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_NSDEV_MAJOR) "-*");
MODULE_ALIAS("/dev/nsdev");
//MODULE_ALIAS("devname:nsdev");
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_NSDEV_MAJOR));
MODULE_ALIAS("/dev/streams/nsdev");
MODULE_ALIAS("/dev/streams/nsdev/*");
MODULE_ALIAS("/dev/streams/clone/nsdev");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

static struct module_info nsdev_minfo = {
	.mi_idnum = CONFIG_STREAMS_NSDEV_MODID,
	.mi_idname = CONFIG_STREAMS_NSDEV_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat nsdev_mstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static struct qinit nsdev_rinit = {
	// qi_putp:putq,
      qi_minfo:&nsdev_minfo,
      qi_mstat:&nsdev_mstat,
};

static struct qinit nsdev_winit = {
	// qi_putp:putq,
      qi_minfo:&nsdev_minfo,
      qi_mstat:&nsdev_mstat,
};

#ifdef CONFIG_STREAMS_NSDEV_MODULE
static
#endif
struct streamtab nsdevinfo = {
      st_rdinit:&nsdev_rinit,
      st_wrinit:&nsdev_winit,
};

/**
 *  nsdevopen:	- open the named streams device
 *  @inode:	shadow special filesystem inode to open
 *  @file:	shadow special filesystem file pointer to open
 *
 *  This is rather simple.  We are going to do a redirected open on the a new device with the major
 *  device number mapped according to name.  We do this by nesting another spec_open() inside the
 *  first one with an adjusted device number. It helps that the orginal file pointer is stored with
 *  the args passed as private_data attached to the current file pointer.  We use this to find the
 *  original file pointer and dentry and get the name of the opened file.
 *
 *  If we don't find a reasonable match and kmod is equipped, we try to load the module with
 *  'streams-' prefixed to the name and run through the list again.  We could also generate the
 *  request from the path to the original dentry.
 */
static int
nsdevopen(struct inode *inode, struct file *file)
{
	struct cdevsw *cdev;
	int err;

	if ((cdev = cdev_match((char *) file->f_dentry->d_name.name))) {
		major_t major = cdev->d_modid;
		minor_t minor = getminor(inode->i_ino);
		dev_t dev = makedevice(major, minor);
		int sflag = (file->f_flags & O_CLONE) ? CLONEOPEN : DRVOPEN;

		err = spec_open(file, cdev, dev, sflag);
		sdev_put(cdev);
	} else
		err = -ENOENT;
	return (err);
}

struct file_operations nsdev_ops ____cacheline_aligned = {
	.owner = NULL,		/* yes NULL */
	.open = nsdevopen,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZATION
 *
 *  -------------------------------------------------------------------------
 */

static struct cdevsw nsdev_cdev = {
	.d_name = CONFIG_STREAMS_NSDEV_NAME,
	.d_str = &nsdevinfo,
	.d_flag = D_MP,
	.d_fop = &nsdev_ops,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  Special open for clone devices.
 *
 *  -------------------------------------------------------------------------
 */

/*
 *  nsdev_open: - open a named clone device node
 *  @inode: the external filesystem inode
 *  @file: the external filesystem file pointer
 *
 *  nsdev_open() is only used to open a named clone device from a character device node in an
 *  external filesystem.  This is never called for direct opens of a specfs device node (for direct
 *  opens see spec_dev_open() in strspecfs.c).  The name of the device in the external filesystem
 *  combined with the minor device number is used to determine the shadow special filesystem
 *  (internal) inode an chain the open call.
 *
 *  This is a separation point between the external and internal filesystem where we convert the
 *  external device number to an internal device number.  The external device number is contained in
 *  inode->i_rdev.
 */

STATIC int
nsdev_open(struct inode *inode, struct file *file)
{
	int err;
	struct cdevsw *cdev;
	major_t major;
	minor_t minor;
	modID_t modid, instance;
	dev_t dev;

#if defined HAVE_KFUNC_TO_KDEV_T
	minor = MINOR(kdev_t_to_nr(inode->i_rdev));
	major = MAJOR(kdev_t_to_nr(inode->i_rdev));
#else
	minor = MINOR(inode->i_rdev);
	major = MAJOR(inode->i_rdev);
#endif
	minor = cdev_minor(&nsdev_cdev, major, minor);
	major = nsdev_cdev.d_major;
	modid = nsdev_cdev.d_modid;
	err = -ENXIO;
	if (!(cdev = cdev_match((char *) file->f_dentry->d_name.name)))
		goto exit;
	err = -ENXIO;
	if (cdev == &nsdev_cdev)
		goto cdev_put_exit;	/* would loop */
	instance = cdev->d_modid;
	dev = makedevice(modid, instance);
	err = spec_open(file, cdev, dev, CLONEOPEN);
      cdev_put_exit:
	sdev_put(cdev);
      exit:
	return (err);
}

STATIC struct file_operations nsdev_f_ops ____cacheline_aligned = {
	.owner = THIS_MODULE,
	.open = nsdev_open,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZAION
 *
 *  -------------------------------------------------------------------------
 */

#ifdef CONFIG_STREAMS_NSDEV_MODULE
static
#endif
int __init
nsdevinit(void)
{
	int err;

#ifdef CONFIG_STREAMS_NSDEV_MODULE
	printk(KERN_INFO NSDEV_BANNER);
#else
	printk(KERN_INFO NSDEV_SPLASH);
#endif
	nsdev_minfo.mi_idnum = modid;
	if ((err = register_cmajor(&nsdev_cdev, major, &nsdev_f_ops)) < 0)
		return (err);
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_NSDEV_MODULE
static
#endif
void __exit
nsdevexit(void)
{
	unregister_cmajor(&nsdev_cdev, major);
};

#ifdef CONFIG_STREAMS_NSDEV_MODULE
module_init(nsdevinit);
module_exit(nsdevexit);
#endif
